consensus/istanbul/backend/handler.go

```diff
@@ -290,3 +298,171 @@ func (sb *Backend) UnregisterPeer(peer consensus.Peer, isProxiedPeer bool) {
 		}
 	}
 }
+
+// Handshake allows this node to identify itself to the peer as a validator and vice versa
+func (sb *Backend) Handshake(peer consensus.Peer) (bool, error) {
+	errCh := make(chan error, 4)
+	isValidatorCh := make(chan bool, 1)
+
+	go func() {
+		validatorProofMessage, err := sb.generateValidatorProofMessage(peer)
+		errCh <- err
+		if err != nil {
+			return
+		}
+		msgBytes, err := validatorProofMessage.Payload()
+		errCh <- err
+		if err != nil {
+			return
+		}
+		errCh <- peer.Send(istanbulValidatorProofMsg, msgBytes)
+	}()
+	go func() {
+		isValidator, err := sb.readValidatorProofMessage(peer)
+		errCh <- err
+		isValidatorCh <- isValidator
+	}()
+
+	timeout := time.NewTimer(handshakeTimeout)
+	defer timeout.Stop()
+	// We write to errCh four times unless if the timeout is triggered or an error isn't nil
+	for i := 0; i < 4; i++ {
+		select {
+		case err := <-errCh:
+			if err != nil {
+				return false, err
+			}
+		case <-timeout.C:
+			return false, p2p.DiscReadTimeout
+		}
+	}
+	return <-isValidatorCh, nil
+}
+
+// generateValidatorProofMessage will create a directAnnounce message that serves
+// as a proof to the peer during the handshake that this node is a validator.
+// If this node is not a validator or this node does not believe the peer
+// is a validator, no proof is generated to hide the address of this node.
+// If this node is a proxy, it will use the most recent directAnnounce this
+// node has received from the proxied validator.
+func (sb *Backend) generateValidatorProofMessage(peer consensus.Peer) (*istanbul.Message, error) {
```

Comment:
nit: I think the naming is a bit confusing here, as `generateValidatorProofMessage` returns the return value of `generateDirectAnnounce`. Consider changing the naming to be more consistent